Sources/SwiftFusion/Inference/PenguinExtensions.swift

```diff
@@ -104,6 +137,577 @@ extension ArrayBuffer: CustomDebugStringConvertible {
+            limitedBy: limit.storage[Type<C.Index>()]
+          ).map(Index.init)
+        },
+        distance: { self_, start, end in
+          self_.storage[unsafelyAssuming: Type<C>()].distance(
+            from: start.storage[Type<C.Index>()], to: end.storage[Type<C.Index>()])
+        },
+        // Hidden collection requirements
+        customIndexOfEquatableElement: { self_, e in
+          self_.storage[unsafelyAssuming: Type<C>()]
+            ._customIndexOfEquatableElement(e).map { $0.map(Index.init) }
+        },
+        customLastIndexOfEquatableElement: { self_, e in
+          self_.storage[unsafelyAssuming: Type<C>()]._customLastIndexOfEquatableElement(e)
+            .map { $0.map(Index.init) }
+        },
+        // MutableCollection requirements.
+        swapAt: { self_, i, j in 
+          self_.storage[unsafelyAssuming: Type<C>()]
+            .swapAt(i.storage[Type<C.Index>()], j.storage[Type<C.Index>()])
+        }
+      )
+    }
+  }
+  
+  public var asAny: Any { storage.asAny }
+}
+
+extension AnyMutableCollection: Sequence {
+  public struct Iterator: IteratorProtocol {
+    var storage: Storage
+    // There's only one operation, so probably no point bothering with a vtable.
+    var next_: (inout Self) -> Element?
+
+    init<T: IteratorProtocol>(_ x: T) where T.Element == Element {
+      storage = .init(x)
+      next_ = { $0.storage[unsafelyAssuming: Type<T>()].next() }
+    }
+
+    public mutating func next() -> Element? { next_(&self) }
+  }
+
+  public func makeIterator() -> Iterator { vtable[0].makeIterator(self) }
+
+  /// If `self` supports a contiguous storage representation, assumes that representation and
+  /// returns the result of passing the storage to `body`; returns `nil` otherwise.
+  ///
+  /// - Note: this documentation might be wrong: https://bugs.swift.org/browse/SR-13486
+  public func withContiguousStorageIfAvailable<R>(
+    _ body: (UnsafeBufferPointer<Element>) throws -> R
+  ) rethrows -> R? {
+    var result: R? = nil
+    try vtable[0].existentialDispatch.withContiguousStorageIfAvailable(storage) {
+      try result = body(
+        UnsafeBufferPointer(start: $0?.assumingMemoryBound(to: Element.self), count: $1))
+    }
+    return result
+  }
+
+  public func _customContainsEquatableElement(_ e: Element) -> Bool? {
+    vtable[0].customContainsEquatableElement(self, e)
+  }
+
+  public __consuming func _copyToContiguousArray() -> ContiguousArray<Element> {
+    vtable[0].copyToContiguousArray(self)
+  }
+  
+  public __consuming func _copyContents(
+    initializing target: UnsafeMutableBufferPointer<Element>
+  ) -> (Iterator,UnsafeMutableBufferPointer<Element>.Index) {
+    vtable[0].copyContents(self, target)
+  }
+}
+
+extension AnyMutableCollection: MutableCollection {
+  public struct Index: Comparable {
+    fileprivate var storage: Storage
+    // Note: using a vtable here actually made benchmarks slower.  We might want to try again if we
```

Comment:
There aren't any benchmarks in this PR. Do you have some benchmarks of this that you have been running?

Reply:
\*sigh\*, you people and your benchmarks! 😉 

The benchmarks for this thing are essentially covered here:
https://github.com/saeta/penguin/blob/master/Sources/Benchmarks/AnyValueBenchmarks.swift#L300-L302
and when this thing becomes public Penguin API they should be expanded.  I don't think benchmarks for this component, which we're going to move to penguin, belong in SwiftFusion.

As for the other parts, I'm not sure I know how to benchmark them.  I'd been assuming that when we get them wired into SwiftFusion we'll observe their performance through existing benchmarks, but if you have other ideas, I'm all ears.